momaland/envs/ingenious/ingenious.py

```diff
@@ -0,0 +1,238 @@
+
+    @override
+    def render(self):
+        """Renders the environment.
+
+        In human mode, it can print to terminal, open
+        up a graphical window, or open up some other display that a human can see and understand.
+        """
+        # self.game.render_game()
+        if self.render_mode is None:
+            warn("You are calling render method without specifying any render mode.")
+            return
+
+    @override
+    def close(self):
+        """Close should release any graphical displays, subprocesses, network connections or any other
+        environment data which should not be kept around after the user is no longer using the environment."""
+        pass
+
+    @override
+    def reset(self, seed=None, options=None):
+        """Reset needs to initialize the `agents` attribute and must set up the environment so that render(),
+        and step() can be called without issues.
+        """
+        if seed is not None:
+            np.random.seed(seed)
+            random.seed(seed)
+        self.game.reset_game(seed)
+        self.agents = self.possible_agents[:]
+        # self.observation_spaces = {agent: self.observe(agent) for agent in self.agents}
+        obs = {agent: self.observe(agent) for agent in self.agents}
+        self.terminations = {agent: False for agent in self.agents}
+        self.truncations = {agent: False for agent in self.agents}
+        self.infos = {agent: {} for agent in self.agents}
+        self.agent_selection = self.agents[self.game.agent_selector]
+        self._cumulative_rewards = {agent: np.zeros(self.num_colors) for agent in self.agents}
+        self.agent_selection = self.agents[self.game.agent_selector]
+        self.rewards = {agent: np.zeros(self.num_colors, dtype="float64") for agent in self.agents}
+        self._cumulative_rewards = {agent: np.zeros(self.num_colors, dtype="float64") for agent in self.agents}
+        self.refresh_cummulative_reward = True
+        return obs, self.infos
+
+    @override
+    def step(self, action):
+        """Steps in the environment.
+
+        Args:
+            action: action of the active agent
+        """
+
+        prev_agent = self.agent_selection
+
+        if self.terminations[prev_agent] or self.truncations[prev_agent]:
+            return self._was_dead_step(action)
+        self.rewards = {agent: np.zeros(self.num_colors, dtype="float64") for agent in self.agents}
+        if self.refresh_cummulative_reward:
+            self._cumulative_rewards[self.agent_selection] = np.zeros(self.num_colors, dtype="float64")
+        if not self.game.end_flag and self.game.return_action_list()[action] == 1:
+            prev_rewards = np.array(list(self.game.score[self.agent_selection].values()))
+            self.game.set_action_index(action)
+            current_rewards = np.array(list(self.game.score[self.agent_selection].values()))
+            self.rewards[prev_agent] = current_rewards - prev_rewards
```

Comment:
I agree with the fact that it is better to keep the checking of rules (and therefore errors) in ingenious_base instead of ingenious, but doesn't the if test here prohibit set_action_index from ingenious_base to throw errors in case of illegal actions?

In set_action_index of ingenious_base you check for valid actions with:
`assert self.masked_action[index] == 1, "Illegal move, choose a valid action."`

However if the action is not a legal action, the condition of the if test here (in ingenious) would already be false due to `self.game.return_action_list()[action] == 1` and therefore never call set_action_index() no?

I tested it with the following code:
```
if __name__ == "__main__":
    env = MOIngenious()
    env.reset()
    # The index() method returns the position at the first occurrence of the specified value.
    # Returns the first illegal action
    illegal_action = env.game.return_action_list().index(0)
    # Returns the first legal action
    legal_action = env.game.return_action_list().index(1)
    # When playing a legal action no error should be thrown
    env.step(legal_action)
    # However when playing an illegal action an error should be thrown
    env.step(illegal_action)
```

When I play an illegal action now, nothing happens (no error, but also no changes in player or environment). I assume that this is because the condition in the if test in the step function is already false and therefore never even runs set_action_index.

Reply:
Updated now to remove the second check in the if statement and differ that check to the assert in the basic class

Reply:
True. Because Ingenious action space is not the same one like in other games, especially for the first round action space. I add more defends in code to assure it robust. But sometimes I am afraid if we add too much assert, it would probably annoy them for they have to understand more rules.
I suppose we leave the code like this so the logic is completed here.